### DISCUSSION TITLE
Cost, scalability and solution architecture.

### DISCUSSION THREAD

**Discussion by ubuku** [ubuku/net.zero#3] [General]
- External adapter design not scalable beyond test net due to economic reason. We will need to request Chainlink to create a pollution data API for the Marketplace.
- To optimise NFT minting fee. Status will be updated every 24 hours, the oracle will still flag every trigger event.
- Using a oracle doesn’t mean decentralisation. The system transmitting that data feed defines the status. Open public source is the ideal data feed of choice for oracles. An oracle by design highlights that systems can have centralised data dependency, which is not necessarily a bad thing.
- Solution architecture should be nailed down, we can be as creative as we like in the test net. We should however consider economic scalability of the solution for main net.
- Minimise on chain computation for gas considerations. Blockchains are best for transaction finality to create trust in a trust less environment. Computation heavy functions should be restricted to logic layer and transaction finality to blockchain layer. In our use case there 1 critical transaction I.e. pollution trigger event record. In the extended use case there are 2 additional events  minting of tax token and  tax settlement.
- Open public data is not the use case, decentralised disclosure of a pollution trigger is the use case. Our source of data is restricted to a single source for proof of concept (POC) however it will included a minimum of 3 independent source to mitigate repeatability bias. Public data satellite data can be easily procured and audited so we don’t have to worry about data authenticity.

**Comment by simonschweizer** [comment]
Hi @ubuku  and @NishithPat 
could you please elaborate on bullet 1 "External adapter design not scalable beyond test net due to economic reason. We will need to request Chainlink to create a pollution data API for the Marketplace." how an external Adapter and the pollution API on node behave differently? I always thought that the available APIs hosted on the market place are in the end also external Adapters. Thx!

**Reply by NishithPat** [reply]
Yeah, you are right. What Kirti is thinking about is having API requests off-chain and only triggering a function on our smart contract if the threshold for a particular parameter is breached. This is to save on gas costs. [External initiators](https://docs.chain.link/docs/external-initiators-introduction/) might be a solution here. But, I do not have an understanding of how to use/build external initiators.

**Reply by ubuku** [reply]
Good question @simonschweizer 
There are a couple of possibilities. 

1. The easy path is do most of the computation off chain and then push the result to a chain link keeper (thats an assumption). which is gas efficient.
2. The second path is to use custom, pure, public functions. these dont utilise any computation gas on EVM and performed calculations/ computations locally.

As @NishithPat rightly quoted we dont have to worry about gas on test net however. It becomes a problem on main-net, we have to optimise the computation formula. I have a clear idea on how to.

The most important question is, what are we computing?  we are computing a specific state update logic which is the secret sauce of the project. I will include the narrative in our project hackMD.
